tracer/src/Datadog.Trace/AppSec/Waf/Waf.cs

```diff
@@ -66,53 +78,107 @@ public void Dispose(bool disposing)
 
             disposed = true;
 
-            rule?.Dispose();
+            wafHandle?.Dispose();
         }
 
-        private static Obj RuleSetFromManifest(List<Obj> argCache)
+        private static Obj CreatObjFromRulesStream(List<Obj> argCache, Stream stream)
         {
-            var assembly = typeof(Waf).Assembly;
-            var resource = assembly.GetManifestResourceStream("Datadog.Trace.AppSec.Waf.rule-set.json");
-            using var reader = new JsonTextReader(new StreamReader(resource));
-            var root = JToken.ReadFrom(reader);
+            using var reader = new StreamReader(stream);
+            using var jsonReader = new JsonTextReader(reader);
+            var root = JToken.ReadFrom(jsonReader);
+
+            if (Log.IsEnabled(LogEventLevel.Debug))
+            {
+                LogRuleDetails(root);
+            }
 
             return Encoder.Encode(root, argCache);
         }
 
-        private static WafHandle NewRule()
+        private static Stream GetRulesManifestStream()
+        {
+            var assembly = typeof(Waf).Assembly;
+            return assembly.GetManifestResourceStream("Datadog.Trace.AppSec.Waf.rule-set.json");
+        }
+
+        private static Stream GetRulesFileStream(string rulesFile)
         {
+            if (!File.Exists(rulesFile))
+            {
+                Log.Error($"AppSec could not find the rules file in path \"{rulesFile}\". AppSec will not run any protections in this application.");
+                return null;
+            }
+
+            return File.OpenRead(rulesFile);
+        }
+
+        // null rulesFile means use rules embedded in the manifest
+        private static WafHandle CreateWafHandle(string rulesFile)
+        {
+            var argCache = new List<Obj>();
+            Obj configObj;
             try
             {
-                DdwafConfigStruct args = default;
+                using var stream = GetRulesStream(rulesFile);
 
-                var argCache = new List<Obj>();
-                var rules = RuleSetFromManifest(argCache);
+                if (stream == null)
+                {
+                    return null;
+                }
 
-                var ruleHandle = WafNative.Init(rules.RawPtr, ref args);
+                configObj = CreatObjFromRulesStream(argCache, stream);
+            }
+            catch (Exception ex)
+            {
+                var message =
+                    rulesFile != null ?
+                    $"AppSec could not read the rule file \"{rulesFile}\" as it was invalid. AppSec will not run any protections in this application." :
+                    "AppSec could not read the rule file emmbeded in the manifest as it was invalid. AppSec will not run any protections in this application.";
```

Comment:
Same comment as above about structured logging and avoiding string allocation if not needed.

Reply:
fixed in 1c0de753d